# Control Flow

## Structural pattern matching with `match`

Starting with Python 3.11, Python has a way to select between multiple options with `match: case` statements.

This is similar to the *case-switch* constructs you find in other programming languages, but more powerful, and also a little bit less predictable, as it can make matches not only on the basis of equal values, but also, in terms of matching types.

 ```python
 match expression:
    case test_expression:
        code_block_for_match
    case _:
        code_block_if_no_match (optional)
 ```

 Usually, the `expression` will be a variable or object, and the `test_expression` can either be a specific value or object, or it can be a type function or class constructor.

 It is also possible in the `test_expression` to match more than one option using the or operator `|`:

In [ ]:
x = 5
match x:
    case "A":
        print("A was selected")
    case str():
        print("Some other string was selected")
    case 0:
        print("Zero was selected")
    case 1 | 2 | 3:
        print(f"the selected value {x} was in the range 1-3")
    case int():
        print(f"An integer other than 0 and 1, 2, and 3 was selected {x}")
    case _:
        print("Neither string nor int")

You can see where the complexities may arise, as an unexpected match can occur and make your code behave unexpectedly.

## The `while` loop

The full while loop in Python is:

```python
while condition:
    body
else:
    post-code
```


When the condition evaluates to `False`, the `while` loop executes the `post-code` section. If the condition is `False` at the beginning, only the `post-code` will be executed.

If a `break` is used within the `body` of the while loop, then the block following the else **will not** be executed.

In [2]:
response = ""
while response != "Q":
    response = input("Q to quit, B to break:")
    if response == "B":
        break
else:
    print("Exited without break")
print("Finished!")


Finished!


The two special statements, `break` and `continue` can be used in the body of a `while` loop to immediately terminate the loop (skipping even the `else` part) or to skip the remainder of the `body` that has not executed yet.

## The `for` loop

In Python, a `for` loop iterates over the values returned by any iterable object. As a result, you can use `for` loops with lists, tuples, string, `range` object and also with generator functions and generator expressions.

```python
for item in sequence:
    body
else:
    post-code
```

The `else` part is similar to the one found in `while` loops. It is executed when no more items in the sequence are left to iterate over, and no `break` statement has been found in the body.

### The `for` loop and tuple unpacking

Python allows you to unpack tuple elements you iterate over with a `for` loop. That makes the syntax cleaner:

In [1]:
tuples = [(1, 2), (3, 7), (9, 5)]
for a, b in tuples:
    print(f"a={a}, b={b}")

a=1, b=2
a=3, b=7
a=9, b=5


### The `enumerate` function

The `enumerate` function takes a sequence and returns both its index and its value. Because there is no `for` to iterate over indices in Python, `enumerate` is quite common:

In [2]:
x = [1, 3, -7, 4, 9, -5, 4]
for i, n in enumerate(x):
    print(f"index={i}, value={n}")

index=0, value=1
index=1, value=3
index=2, value=-7
index=3, value=4
index=4, value=9
index=5, value=-5
index=6, value=4


### The `zip` function

Sometimes its useful to combine two or more iterables together before looping over them. 

The `zip` function does just that: combines the elements from one or more iterables into a tuple until it reaches the end of the **shortest** iterable:

In [4]:
x = [1, 2, 3, 4]
y = ["a", "b", "c"]

x = zip(x, y)
assert list(x) == [(1, "a"), (2, "b"), (3, "c")]

### List, set, and dictionary comprehensions

In Python, common situation in which you have a `for` loop to iterate through a sequence, modifying or selecting individual elements from the sequence, to finally create a new list is supported by a special syntax construct called a **comprehension**.

You can think of a list, set, or dictionary comprehensions as a one-line `for` loop that creates a new list, set or dictionary from a sequence.

```python
# list comprehension
new_list = [expressin1 for variable in old_sequence if expression]

# set comprehension
new_set = {expressin1 for variable in old_sequence if expression}

# dictionary comprehension
new_dict = {k:v for k, v in old_sequence if expression}
```

In [6]:
x = range(1, 5 + 1)
x_squared = [n * n for n in x]
assert x_squared == [1, 4, 9, 16, 25]

x_even_squared = [n * n for n in x if n % 2 == 0]
assert x_even_squared == [4, 16]

x_squared_dict = {n: n * n for n in x}
assert x_squared_dict == {1: 1, 2: 4, 3: 9, 4: 16, 5: 25}

### Generator expressions

A generator expression is similar to a list comprehension. It uses parentheses, instead of square brackets, but the syntax is the same:

In [7]:
x = [1, 2, 3, 4, 5]
x_squared = (n * n for n in x)
assert list(x_squared) == [1, 4, 9, 16, 25]

Generator expressions can be iterated over as any other iterable:

In [8]:
x = [1, 2, 3, 4, 5]
x_squared = (n * n for n in x)

for n in x_squared:
    print(n)

1
4
9
16
25


The advantage of using a generator expression is that the entire list of entries are not materialized in memory, so arbitrarily large sequences can be generated with very little memory overhead.

## Breaking up statements across multiple lines: indentation considerations

In Python, you can explicitly break up a line by using the backslash character `\`.

In [17]:
# This won't work
# x = 100 + 200 + 300
#     + 400 + 500

# this works
x = 100 + 200 + 300 \
    + 400 + 500

# this works too
x = 100 + 200 + 300 \
+ 400 + 500


You can break up strings by `\` as well, but any indentation tabs or spaces become part of the string, which might not be what you intended:

In [18]:
s = "a very large string that most probably will reach the threshold that I had \
    established"

print(s)

a very large string that most probably will reach the threshold that I had     established


It's quite common to break such strings with the help of parentheses `()`

In [19]:
s = ("a very large string that most probably will reach the threshold that I "
     "had established")
print(s)

a very large string that most probably will reach the threshold that I had established


Alternatively, you can also use `"` and the backslash:

In [20]:
s = "a very large string that most probably will reach the threshold that I " \
    "had established"
print(s)

a very large string that most probably will reach the threshold that I had established


## Boolean values

Most Python objects can be used as Booleans. This is the recommended way to use them as this makes the syntax more succinct:

+ The numbers `0`, `0.0`, and `0+0j` are all `False`. Any other number is `True`.
+ The empty string `""` is `False`; any other string is `True`.
+ The empty list `[]` is `False`; any other list is `True`.
+ The empty dictionary `{}` is `False`; any other dictionary is `True`.
+ The empty set `set()` is `False`; any other set is `True`.
+ The special value `None` is alwasy `False`.

### Comparing with `==` and `!=` and `is` and `is not`

The `==` and `!=` are the operators used in most situations to test if their operands contain the same values.

By contrast, `is` and `is not` test whether their operands are the **same object**:

In [23]:
x = [1, 2]
y = [1, 2]
assert x == y
assert x is not y
assert x[0] is y[0]  # Because ints are immutable

### Exercise

Create a Python utility that replicates the UNIX `wc` utility that reports the number of lines, words, and characters in a file.

Let's start by trying out what `wc` returns when applied to our Moby Dick file:

In [24]:
%%bash
wc data/moby_01.txt

  26  273 1509 data/moby_01.txt


This means the file has 26 lines, 273 words, and 1509 characters.

Let's create a Python snippet that does the same:

In [25]:
filename = "data/moby_01.txt"
num_lines = 0
num_words = 0
num_chars = 0
with open(filename, "r") as infile:
    for line in infile:
        num_lines += 1
        num_words += len(line.split())
        num_chars += len(line)

print(f"Lines: {num_lines}, Words: {num_words}, Characters: {num_chars}")

Lines: 26, Words: 273, Characters: 1509


In [26]:
filename = "data/word_count.tst"
num_lines = 0
num_words = 0
num_chars = 0
with open(filename, "r") as infile:
    for line in infile:
        num_lines += 1
        num_words += len(line.split())
        num_chars += len(line)

print(f"Lines: {num_lines}, Words: {num_words}, Characters: {num_chars}")

Lines: 4, Words: 30, Characters: 189
